### PR TITLE
Allow to dial any angle with digital joystick

### PR DESCRIPTION
**Pull request by german77**
Instead of returning only the cardinal angles with digital buttons allow the input to output any angle by drifting the angle slowly to the desired direction. Solving #4580, #2845. A solution with mouse will also be implemented in a future PR.

The speed of the drift can be adjusted in the config.ini. If necessary I can add the setting to the UI but as far my testing with the game I have needs no adjustment.

### CODE REVIEW

File: src/input_common/analog_from_button.cpp [hunk 1]
```diff
@@ -59,6 +134,11 @@ class Analog final : public Input::AnalogDevice {
     Button right;
     Button modifier;
     float modifier_scale;
+    float modifier_angle;
+    float angle = 0;
+    float amplitude = 0;
+    std::thread update_thread;
+    bool update_thread_running = true;
```
Comment (Morph1984) [inline src/input_common/analog_from_button.cpp:141]: Braced initializers
```suggestion
    float angle{};
    float amplitude{};
    std::thread update_thread;
    bool update_thread_running{true};
```

File: src/input_common/analog_from_button.cpp [hunk 2]
```diff
@@ -11,31 +14,103 @@ class Analog final : public Input::AnalogDevice {
...
         }
-        if (left->GetStatus()) {
-            --x;
-        }
-        if (up->GetStatus()) {
-            ++y;
+    }
+
+    void MoveToDirection(bool enable, float to_angle) {
+        constexpr float TAU = 3.1415926f * 2.0f;
```
Comment (Morph1984) [inline src/input_common/analog_from_button.cpp:32]: make use of the Common::PI constant in `common/math_util.h`
```suggestion
        constexpr float TAU = Common::PI * 2.0f;
```

File: src/input_common/analog_from_button.cpp [hunk 3]
```diff
@@ -59,6 +134,11 @@ class Analog final : public Input::AnalogDevice {
     Button right;
     Button modifier;
     float modifier_scale;
+    float modifier_angle;
+    float angle{0};
+    float amplitude{0};
```
Comment (Morph1984) [inline src/input_common/analog_from_button.cpp:139]: The default braced initialization for `float example_float{}` is 0.0f

File: src/input_common/analog_from_button.cpp [hunk 4]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            }
+            if (u && d) {
+                u = false;
+                d = false;
+            }
+
+            // Move to the right
+            MoveToDirection(r & !u & !d, 0.0f);
+
+            // Move to the right up
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:79]:
```suggestion
            // Move to the upper right
```

File: src/input_common/analog_from_button.cpp [hunk 5]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            MoveToDirection(r & u & !d, Common::PI * 0.25f);
+
+            // Move up
+            MoveToDirection(u & !l & !r, Common::PI * 0.5f);
+
+            // Move to the left up
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:85]:
```suggestion
            // Move to the upper left
```

File: src/input_common/analog_from_button.cpp [hunk 6]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            MoveToDirection(l & u & !d, Common::PI * 0.75f);
+
+            // Move to the left
+            MoveToDirection(l & !u & !d, Common::PI);
+
+            // Move to the left down
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:91]:
```suggestion
            // Move to the bottom left
```

File: src/input_common/analog_from_button.cpp [hunk 7]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            MoveToDirection(l & !u & d, Common::PI * 1.25f);
+
+            // Move down
+            MoveToDirection(d & !l & !r, Common::PI * 1.5f);
+
+            // Move to the right down
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:97]:
```suggestion
            // Move to the bottom right
```

File: src/input_common/analog_from_button.cpp [hunk 8]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+                r = false;
+                l = false;
+            }
+            if (u && d) {
+                u = false;
+                d = false;
+            }
+
+            // Move to the right
+            MoveToDirection(r & !u & !d, 0.0f);
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:77]: Note that using bitwise arithmetic on booleans can cause warnings and also be flagged by static analysis tooling. You may want to consider using the logical operators instead.

File: src/input_common/analog_from_button.cpp [hunk 9]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
         }
+    }
 
-        const float coef = modifier->GetStatus() ? modifier_scale : 1.0f;
-        return std::make_tuple(static_cast<float>(x) * coef * (y == 0 ? 1.0f : SQRT_HALF),
-                               static_cast<float>(y) * coef * (x == 0 ? 1.0f : SQRT_HALF));
+    std::tuple<float, float> GetStatus() const override {
+        return std::make_tuple(cos(angle) * amplitude, sin(angle) * amplitude);
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:111]:
```suggestion
        return std::make_tuple(std::cos(angle) * amplitude, std::sin(angle) * amplitude);
```

File: src/input_common/analog_from_button.cpp [hunk 10]
```diff
@@ -11,31 +14,101 @@ class Analog final : public Input::AnalogDevice {
...
+            // Move to the right down
+            MoveToDirection(r & !u & d, Common::PI * 1.75f);
+
+            // Move if a key is pressed
+            if (r | l | u | d) {
+                amplitude = coef;
+            } else {
+                amplitude = 0;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:106]: There should probably be a comment here indicating why the sleep is necessary.

File: src/input_common/analog_from_button.cpp (hunk 9 above)
Comment (Morph1984) [reply src/input_common/analog_from_button.cpp:111]: Would a `<cmath>` include be needed as well?

File: src/input_common/analog_from_button.cpp [hunk 11]
```diff
@@ -11,31 +15,103 @@ class Analog final : public Input::AnalogDevice {
...
-        }
-        if (up->GetStatus()) {
-            ++y;
+    }
+
+    void MoveToDirection(bool enable, float to_angle) {
+        constexpr float TAU = Common::PI * 2.0f;
+        // Use wider angle to ease the transition.
+        constexpr float aperture = TAU * 0.15f;
+        float top_limit = to_angle + aperture;
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:36]: Missed in the initial review, but `top_limit` and `bottom_limit` can be const.

File: src/input_common/analog_from_button.cpp [hunk 12]
```diff
@@ -11,31 +15,103 @@ class Analog final : public Input::AnalogDevice {
...
+        float bottom_limit = to_angle - aperture;
+
+        if (enable) {
```
Comment (lioncash) [inline src/input_common/analog_from_button.cpp:39]:
```suggestion
        if (!enable) {
            return;
        }
```

Can likely be placed at the top of the function, unindenting all the contained code here.

**Review by lioncash — COMMENTED** [review]
Just two things I missed in the previous review. Other than those it looks good

**Review by bunnei — APPROVED** [review]

**Pull request merged**